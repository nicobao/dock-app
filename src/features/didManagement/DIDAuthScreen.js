import {addTestId} from 'src/core/automation-utils';
import React, {useState} from 'react';
import {
  Box,
  Header,
  NBox,
  ScreenContainer,
  Typography,
  Input,
} from '../../design-system';
import {ScrollView} from 'react-native';
import {Button, Spinner, Stack, FormControl} from 'native-base';
import {translate} from '../../locales';
import {navigate} from '../../core/navigation';
import {Routes} from '../../core/routes';
import {validateEmail, getScopeFields, extractClientInfo} from './didAuthUtils';
import {KeyboardAvoidingView, Platform} from 'react-native';
import {CustomSelectInput} from '../../components/CustomSelectInput';
import {useDIDAuth, useDIDAuthHandlers} from './didAuthHooks';

export function DIDAuthConfirmScreen({
  authenticateDID,
  profileData,
  selectedDID,
  clientInfo,
  dids = [],
  handleChange,
}) {
  const [error, setError] = useState();
  const {name, scope} = clientInfo;
  const fields = getScopeFields(scope);

  function handleCancel() {
    navigate(Routes.ACCOUNTS);
  }

  function handleSubmit() {
    for (let i = 0; i < fields.length; i++) {
      const field = fields[i];
      const value = profileData[field.id];
      if (field.required && !value) {
        setError(`${field.name} is required`);
        return;
      }

      if (value && field.type === 'email' && !validateEmail(value)) {
        setError(`${field.name} must be a valid email address`);
        return;
      }
    }

    setError(null);
    authenticateDID();
  }

  return (
    <ScreenContainer testID="DIDAuthConfirmScreen">
      <ScrollView
        contentContainerStyle={{flexGrow: 1}}
        keyboardShouldPersistTaps="handled">
        <Header marginTop={22}>
          <Box
            marginLeft={22}
            marginRight={22}
            flexDirection="row"
            alignItems="center">
            <Box flex={1}>
              <Typography variant="h1" textAlign="center">
                {translate('auth.authorize')} {name}
              </Typography>
            </Box>
          </Box>
        </Header>
        <Stack direction={'column'} mx={4}>
          <NBox mb={50}>
            <Typography textAlign="center" variant="list-description">
              {translate('auth.sign_in_msg')}
            </Typography>
          </NBox>

          <FormControl isInvalid={!!error}>
            <Stack mt={7}>
              <FormControl.Label>
                {translate('didManagement.did_name')}
              </FormControl.Label>
              <CustomSelectInput
                {...addTestId('DID')}
                onPressItem={item => {
                  handleChange('did', item.value);
                }}
                renderItem={item => {
                  return (
                    <>
                      <Typography
                        numberOfLines={1}
                        textAlign="left"
                        variant="description">
                        {item.label}
                      </Typography>
                      <Typography
                        numberOfLines={1}
                        textAlign="left"
                        variant="screen-description">
                        {item.description}
                      </Typography>
                    </>
                  );
                }}
                items={dids}
              />
            </Stack>
          </FormControl>

          <FormControl isInvalid={!!error}>
            <Box marginTop={12}>
              {fields.map(field => (
                <>
                  <Typography variant="field-label">{field.name}:</Typography>
                  <Box marginTop={4} marginBottom={12}>
                    <Input
                      placeholder={field.placeholder}
                      {...addTestId(field.name)}
                      value={profileData[field.id]}
                      onChangeText={value => {
                        handleChange(field.id, value);
                      }}
                      autoCapitalize="none"
                      secureTextEntry={field.type === 'password'}
                    />
                  </Box>
                </>
              ))}
            </Box>
            <FormControl.ErrorMessage>{error}</FormControl.ErrorMessage>
          </FormControl>
        </Stack>

        <Stack direction={'row'} mt={'auto'} mx={4} mb={10}>
          <Button.Group
            mx={{
              base: 'auto',
              md: 0,
            }}
            size="md">
            <Button onPress={handleCancel} width="48.5%" colorScheme="dark">
              <Typography>{translate('navigation.cancel')}</Typography>
            </Button>
            <Button
              onPress={handleSubmit}
              disabled={!selectedDID}
              width="48.5%">
              <Typography>{translate('navigation.approve')}</Typography>
            </Button>
          </Button.Group>
        </Stack>
      </ScrollView>
    </ScreenContainer>
  );
}

export function DIDAuthScreen({authState, retry}) {
  return (
    <ScreenContainer testID="DIDAuthScreen">
      <Header>
        <Box
          marginLeft={22}
          marginRight={22}
          flexDirection="row"
          alignItems="center">
          <Box flex={1}>
            <Typography fontFamily="Montserrat" fontSize={24} fontWeight="600">
              {translate('qr_scanner.header')}
            </Typography>
          </Box>
        </Box>
      </Header>
      <Stack direction={'column'} mx={70} mt={40}>
        <NBox mt={15}>
          {authState === 'processing' ? (
            <Spinner />
          ) : (
            <Typography
              style={{
                textAlign: 'center',
              }}
              variant="list-description">
              {translate(`qr_scanner.${authState}`)}
            </Typography>
          )}
        </NBox>
        {authState === 'error' ? (
          <Button onPress={retry} mt={7}>
            <Typography>{translate('qr_scanner.retry')}</Typography>
          </Button>
        ) : null}
      </Stack>
    </ScreenContainer>
  );
}

export function DIDAuthScreenContainer({route}) {
  const {dockWalletAuthDeepLink} = route.params || {};

  const {dids, authState, handleRetry, authenticateDID} = useDIDAuth();
  const {profileData, selectedDID, handleChange} = useDIDAuthHandlers();
  const clientInfo = extractClientInfo(dockWalletAuthDeepLink);

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={{flex: 1}}>
      {authState === 'start' ? (
        <DIDAuthConfirmScreen
          {...{
            authenticateDID: () => {
              authenticateDID({
                dockWalletAuthDeepLink,
                selectedDID,
                profileData,
              });
            },
            profileData,
            dids,
            selectedDID,
            clientInfo,
            handleChange,
          }}
        />
      ) : (
        <DIDAuthScreen authState={authState} retry={handleRetry} />
      )}
    </KeyboardAvoidingView>
  );
}
