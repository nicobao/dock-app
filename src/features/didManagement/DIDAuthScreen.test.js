import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {DIDAuthScreen, DIDAuthConfirmScreen} from './DIDAuthScreen';

const mockStore = configureMockStore();

describe('QRScanScreen', () => {
  it('should render DIDAuthScreen correctly ', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(<DIDAuthScreen authState={'start'} />, {
      context: {store: mockStore(initialState)},
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('should render DIDAuthConfirmScreen ', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(
      <DIDAuthConfirmScreen
        {...{
          authenticateDID: jest.fn(),
          profileData: {},
          dids: [],
          selectedDID: null,
          clientInfo: {
            name: 'Unnamed App',
            website: 'test.com',
            scope: 'public',
          },
          handleChange: jest.fn(),
        }}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
