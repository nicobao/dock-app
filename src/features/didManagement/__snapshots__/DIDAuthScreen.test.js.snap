// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRScanScreen should render DIDAuthConfirmScreen  1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="DIDAuthConfirmScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <ScrollView
        contentContainerStyle={
          Object {
            "flexGrow": 1,
          }
        }
        keyboardShouldPersistTaps="handled"
      >
        <Styled(View)
          marginTop={22}
        >
          <Box
            alignItems="center"
            flexDirection="row"
            marginLeft={22}
            marginRight={22}
          >
            <Box
              flex={1}
            >
              <Styled(Component)
                textAlign="center"
                variant="h1"
              >
                Authorize
                 
                Unnamed App
              </Styled(Component)>
            </Box>
          </Box>
        </Styled(View)>
        <Memo(ForwardRef(Stack))
          direction="column"
          mx={4}
        >
          <Memo(ForwardRef(Box))
            mb={50}
          >
            <Styled(Component)
              textAlign="center"
              variant="list-description"
            >
              Sign in with your Web3 ID
            </Styled(Component)>
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(FormControl))
            isInvalid={false}
          >
            <Memo(ForwardRef(Stack))
              mt={7}
            >
              <Memo(ForwardRef(FormControlLabel))>
                DID Name
              </Memo(ForwardRef(FormControlLabel))>
              <CustomSelectInput
                accessibilityLabel="DID"
                items={Array []}
                onPressItem={[Function]}
                renderItem={[Function]}
                testID="DID"
              />
            </Memo(ForwardRef(Stack))>
          </Memo(ForwardRef(FormControl))>
          <Memo(ForwardRef(FormControl))
            isInvalid={false}
          >
            <Box
              marginTop={12}
            >
              <Styled(Component)
                variant="field-label"
              >
                Name
                :
              </Styled(Component)>
              <Box
                marginBottom={12}
                marginTop={4}
              >
                <Memo(ForwardRef(Input))
                  accessibilityLabel="Name"
                  autoCapitalize="none"
                  onChangeText={[Function]}
                  placeholder="John Doe"
                  secureTextEntry={false}
                  testID="Name"
                />
              </Box>
            </Box>
            <Memo(ForwardRef(FormControlErrorMessage)) />
          </Memo(ForwardRef(FormControl))>
        </Memo(ForwardRef(Stack))>
        <Memo(ForwardRef(Stack))
          direction="row"
          mb={10}
          mt="auto"
          mx={4}
        >
          <Memo(ForwardRef)
            mx={
              Object {
                "base": "auto",
                "md": 0,
              }
            }
            size="md"
          >
            <Memo(ForwardRef(Button))
              colorScheme="dark"
              onPress={[Function]}
              width="48.5%"
            >
              <Styled(Component)>
                Cancel
              </Styled(Component)>
            </Memo(ForwardRef(Button))>
            <Memo(ForwardRef(Button))
              disabled={true}
              onPress={[Function]}
              width="48.5%"
            >
              <Styled(Component)>
                Approve
              </Styled(Component)>
            </Memo(ForwardRef(Button))>
          </Memo(ForwardRef)>
        </Memo(ForwardRef(Stack))>
      </ScrollView>
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;

exports[`QRScanScreen should render DIDAuthScreen correctly  1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="DIDAuthScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={22}
          marginRight={22}
        >
          <Box
            flex={1}
          >
            <Styled(Component)
              fontFamily="Montserrat"
              fontSize={24}
              fontWeight="600"
            >
              Web3 ID Authorization
            </Styled(Component)>
          </Box>
        </Box>
      </Styled(View)>
      <Memo(ForwardRef(Stack))
        direction="column"
        mt={40}
        mx={70}
      >
        <Memo(ForwardRef(Box))
          mt={15}
        >
          <Styled(Component)
            style={
              Object {
                "textAlign": "center",
              }
            }
            variant="list-description"
          >
            Start
          </Styled(Component)>
        </Memo(ForwardRef(Box))>
      </Memo(ForwardRef(Stack))>
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;
